onyo/lib/onyo.py

```diff
@@ -1145,54 +1145,53 @@ def _read_asset(self, asset: Path) -> dict:
     def _set_sanitize(self, paths: Iterable[Union[Path, str]],
                       depth: Union[int, None]) -> list[Path]:
         """
-        Check and normalize a list of paths. Traverse directories and select all
-        assets from `paths` and return them as a list of absolute Paths.
+        Check and normalize a list of paths. Select all assets in the
+        repository that are relative to the given `paths` descending at most
+        `depth` directories. A `depth` of 0 descends without a limit.
         """
-        error_path_absent = []
-        error_path_protected = []
-        paths_to_set = []
-
         if depth and depth < 0:
             log.error(f"depth values must be positive, but is {depth}.")
             raise ValueError(f"depth values must be positive, but is {depth}.")
 
-        for p in paths:
-            full_path = Path(self.opdir, p).resolve()
-
-            # paths must exist
-            if not full_path.exists():
-                error_path_absent.append(p)
-                continue
-
-            # protected paths
-            if self._is_protected_path(full_path):
-                error_path_protected.append(p)
-                continue
-
-            if full_path.is_dir():
-                paths_to_set += self._select_assets_from_directory(full_path, depth)
-            else:
-                paths_to_set.append(full_path)
+        paths = {Path(p) for p in paths}
+
+        # TODO: This is superfluous due to using self.assets to find valid
+        #   set/unset targets -- it'll never hit non-existing or protected
+        #   paths
+        invalid_paths = {p for p in paths if not p.exists()}
+        if any(invalid_paths):
+            msg = (
+                'The following paths do not exist:\n{0}\n'
+                'Nothing was deleted.').format(
+                    '\n'.join([str(p) for p in invalid_paths]))
+            log.error(msg)
+            raise FileNotFoundError(msg)
+
+        # TODO: This is likewise superfluous
+        protected_paths = {p for p in paths if self._is_protected_path(p)}
+        if any(protected_paths):
+            msg = (
+                'The following paths are protected by onyo:\n{0}\n'
+                'Nothing was deleted').format(
+                    '\n'.join([str(p) for p in protected_paths]))
+            log.error(msg)
+            raise OnyoProtectedPathError(msg)
+
+        # Select assets that match paths and depth
+        #  note: given paths that are asset paths (rather than directories)
+        #  only select the asset if it appears in self.assets
+        assets = [
+            a for a in self.assets if any([
```

Comment:
What does `any` do here? I've read this a few times now, and I can't figure out what it's adding to the conditionals already present.

Reply:
It checks if an asset if relative to and meets the depth requirement to any of the given paths. For example: Lets say we have an asset `foo/bar/baz/laptop.1`, and we enter the paths `['abc/def', 'foo/bar', 'lorem/ipsum']` with a depth of 2. If we check whether this asset meets our requirements, we get `[False, True, False]`. Our paths are inclusive rather than restrictive, so even one `True` means we want to include the asset, hence the use of `any`.

Reply:
Ahh; now I see the purpose of `any()` here. All good, and thanks for explaining. :-)